File: lib/trash.ts
import fs from 'fs/promises';
import path from 'path';
import crypto from 'crypto';

const WIKI_DATA_DIR = path.join(process.cwd(), 'wiki-data');
const TRASH_DIR = path.join(WIKI_DATA_DIR, '.trash');
const TRASH_META = path.join(TRASH_DIR, 'meta.json');

export interface TrashItem {
  id: string;
  originalPath: string;
  isFolder: boolean;
  deletedAt: string;
  name: string;
}

// 进程级锁
let lockPromise: Promise<void> = Promise.resolve();
function withLock<T>(fn: () => Promise<T>): Promise<T> {
  const prev = lockPromise;
  let resolve: () => void;
  lockPromise = new Promise<void>((r) => { resolve = r; });
  return prev.then(fn).finally(() => resolve!());
}

async function ensureTrashDir() {
  await fs.mkdir(TRASH_DIR, { recursive: true });
}

async function readMeta(): Promise<TrashItem[]> {
  try {
    const raw = await fs.readFile(TRASH_META, 'utf-8');
    return JSON.parse(raw);
  } catch {
    return [];
  }
}

async function writeMeta(items: TrashItem[]) {
  await fs.writeFile(TRASH_META, JSON.stringify(items, null, 2), 'utf-8');
}

/**
 * 将文件/文件夹移入回收站
 */
export async function moveToTrash(articlePath: string, isFolder: boolean): Promise<TrashItem> {
  return withLock(async () => {
    await ensureTrashDir();

    const id = crypto.randomBytes(8).toString('hex');
    const name = articlePath.split('/').pop() || articlePath;

    // 源路径
    let sourcePath: string;
    if (isFolder) {
      sourcePath = path.join(WIKI_DATA_DIR, articlePath);
    } else {
      sourcePath = path.join(WIKI_DATA_DIR, articlePath + '.md');
      // 可能是父页面
      try { await fs.access(sourcePath); } catch {
        sourcePath = path.join(WIKI_DATA_DIR, articlePath, '_index.md');
      }
    }

    // 目标路径
    const destPath = path.join(TRASH_DIR, id);

    // 移动
    await fs.rename(sourcePath, destPath);

    const item: TrashItem = {
      id,
      originalPath: articlePath,
      isFolder,
      deletedAt: new Date().toISOString(),
      name,
    };

    const meta = await readMeta();
    meta.unshift(item);
    await writeMeta(meta);

    return item;
  });
}

/**
 * 从回收站恢复
 */
export async function restoreFromTrash(trashId: string): Promise<boolean> {
  return withLock(async () => {
    const meta = await readMeta();
    const idx = meta.findIndex((i) => i.id === trashId);
    if (idx === -1) return false;

    const item = meta[idx];
    const trashPath = path.join(TRASH_DIR, item.id);

    // 恢复到原始路径
    let destPath: string;
    if (item.isFolder) {
      destPath = path.join(WIKI_DATA_DIR, item.originalPath);
    } else {
      destPath = path.join(WIKI_DATA_DIR, item.originalPath + '.md');
    }

    // 确保父目录存在
    await fs.mkdir(path.dirname(destPath), { recursive: true });

    // 检查目标是否已存在
    try {
      await fs.access(destPath);
      return false; // 目标已存在，不覆盖
    } catch {
      // 目标不存在，可以安全恢复
    }

    await fs.rename(trashPath, destPath);

    meta.splice(idx, 1);
    await writeMeta(meta);
    return true;
  });
}

/**
 * 永久删除回收站中的项目
 */
export async function permanentDelete(trashId: string): Promise<boolean> {
  return withLock(async () => {
    const meta = await readMeta();
    const idx = meta.findIndex((i) => i.id === trashId);
    if (idx === -1) return false;

    const trashPath = path.join(TRASH_DIR, meta[idx].id);
    try {
      const stat = await fs.stat(trashPath);
      if (stat.isDirectory()) {
        await fs.rm(trashPath, { recursive: true });
      } else {
        await fs.unlink(trashPath);
      }
    } catch { /* already gone */ }

    meta.splice(idx, 1);
    await writeMeta(meta);
    return true;
  });
}

/**
 * 获取回收站列表
 */
export async function listTrash(): Promise<TrashItem[]> {
  await ensureTrashDir();
  return readMeta();
}

/**
 * 清空回收站
 */
export async function emptyTrash(): Promise<void> {
  return withLock(async () => {
    const meta = await readMeta();
    for (const item of meta) {
      const trashPath = path.join(TRASH_DIR, item.id);
      try {
        const stat = await fs.stat(trashPath);
        if (stat.isDirectory()) {
          await fs.rm(trashPath, { recursive: true });
        } else {
          await fs.unlink(trashPath);
        }
      } catch { /* skip */ }
    }
    await writeMeta([]);
  });
}

File: backend/lib/trash.py
import asyncio
import json
import os
import shutil
from datetime import datetime, timezone
from pathlib import Path

from backend.config import WIKI_DATA_DIR, TRASH_DIR

_lock = asyncio.Lock()
_TRASH_META = TRASH_DIR / "meta.json"


async def _ensure_trash_dir() -> None:
    await asyncio.to_thread(TRASH_DIR.mkdir, parents=True, exist_ok=True)


async def _read_meta() -> list[dict]:
    try:
        raw = await asyncio.to_thread(lambda: _TRASH_META.read_text("utf-8"))
        return json.loads(raw)
    except (FileNotFoundError, json.JSONDecodeError):
        return []


async def _write_meta(items: list[dict]) -> None:
    data = json.dumps(items, ensure_ascii=False, indent=2)
    await asyncio.to_thread(lambda: _TRASH_META.write_text(data, "utf-8"))


async def move_to_trash(article_path: str, is_folder: bool) -> dict:
    async with _lock:
        await _ensure_trash_dir()

        trash_id = os.urandom(8).hex()
        name = article_path.split("/")[-1] or article_path

        if is_folder:
            source = WIKI_DATA_DIR / article_path
        else:
            source = WIKI_DATA_DIR / (article_path + ".md")
            if not source.exists():
                source = WIKI_DATA_DIR / article_path / "_index.md"

        dest = TRASH_DIR / trash_id
        await asyncio.to_thread(source.rename, dest)

        item = {
            "id": trash_id,
            "originalPath": article_path,
            "isFolder": is_folder,
            "deletedAt": datetime.now(timezone.utc).isoformat(),
            "name": name,
        }

        meta = await _read_meta()
        meta.insert(0, item)
        await _write_meta(meta)
        return item


async def restore_from_trash(trash_id: str) -> bool:
    async with _lock:
        meta = await _read_meta()
        idx = next((i for i, m in enumerate(meta) if m["id"] == trash_id), -1)
        if idx == -1:
            return False

        item = meta[idx]
        trash_path = TRASH_DIR / item["id"]

        if item["isFolder"]:
            dest = WIKI_DATA_DIR / item["originalPath"]
        else:
            dest = WIKI_DATA_DIR / (item["originalPath"] + ".md")

        await asyncio.to_thread(dest.parent.mkdir, parents=True, exist_ok=True)

        if dest.exists():
            return False

        await asyncio.to_thread(trash_path.rename, dest)
        meta.pop(idx)
        await _write_meta(meta)
        return True


async def permanent_delete(trash_id: str) -> bool:
    async with _lock:
        meta = await _read_meta()
        idx = next((i for i, m in enumerate(meta) if m["id"] == trash_id), -1)
        if idx == -1:
            return False

        trash_path = TRASH_DIR / meta[idx]["id"]
        try:
            if trash_path.is_dir():
                await asyncio.to_thread(shutil.rmtree, str(trash_path))
            else:
                await asyncio.to_thread(trash_path.unlink)
        except FileNotFoundError:
            pass

        meta.pop(idx)
        await _write_meta(meta)
        return True


async def list_trash() -> list[dict]:
    await _ensure_trash_dir()
    return await _read_meta()


async def empty_trash() -> None:
    async with _lock:
        meta = await _read_meta()
        for item in meta:
            trash_path = TRASH_DIR / item["id"]
            try:
                if trash_path.is_dir():
                    await asyncio.to_thread(shutil.rmtree, str(trash_path))
                else:
                    await asyncio.to_thread(trash_path.unlink)
            except FileNotFoundError:
                pass
        await _write_meta([])
